content/2017/02/06/server-side-rendering-client-side-app.rst

```diff
@@ -0,0 +1,257 @@
+
+The second and third options are similar. Using a ``<script>`` tag is probably
+the most efficient (requiring only one DOM interaction to acquire the entire
+data set), but requires careful namespacing and patterns for which data is
+attached to which existing markup. In cases involving large collections, this
+would be my preferred approach.
+
+Storing JSON in data-attributes on individual DOM elements has the advantage of
+coupling the data and markup together for each component, but requires
+consistent markup patterns if the JS is to be reusable for various pieces of
+the app. It necessitates DOM interactions to fetch the data, which could easily
+cause performance issues with larger collections. In our case — with a
+relatively small data set for each page — this option seemed to provide both
+reasonable performance and a clear relationship between the data and its
+corresponding markup.
+
+For example, let's say that we want to attach models and views to a
+server-rendered list of comments. Using `Jinja2`_/`Nunjucks`_, our markup might
+look like this:
+
+.. code:: jinja
+
+    <div class="comment-list">
+      <article class="comment" data-js-model="{{ comment|json }}">
+        <p>{{ comment.body }}</p>
+        <p>{{ comment.author }}</p>
+      </article>
+    </div>
+
+.. _Jinja2: http://jinja.pocoo.org/docs/dev/
+.. _Nunjucks: https://mozilla.github.io/nunjucks/
+
+
+Brief ``<aside>``
+-----------------
+
+You'll note that this markup is using a custom ``json`` filter to convert an
+object into a JSON string. One downside of sharing templates between front-end
+and back-end (Nunjucks written in JS on the front-end, and Jinja2 written in
+Python on the back-end) is that any custom filters used in shared templates
+must be written in both languages. So for this to work, we have a ``json``
+filter added to our Nunjucks environment:
+
+.. code:: js
+
+    import nunjucks from 'nunjucks';
+
+    const env = new nunjucks.Environment();
+    env.addFilter('json', (val) => JSON.stringify(val));
+
+And a corresponding filter added to our Jinja2 environment:
+
+.. code:: python
+
+    from json import dumps
+    from jinja2 import Environment
+
+    def environment(**options):
+        env = Environment(**options)
+        env.filters.update({
+            'json': json,
+        })
+        return env
+
+    def json(val):
+        """Return given value as a JSON string."""
+        return dumps(val)
+
+This isn't ideal, but seems like a reasonable trade-off since it allows us to
+avoid duplicating all the various template files.
+
+Ok, ``</aside>``.
+
+
+Using the Data
+--------------
+
+So we've made the model/collection data available in the DOM without requiring
+an additional XHR. Now we need to add our JS layer, turning the data into
+actual models or collections managed by views.
+
+The details differ here from one framework to another. Since we're using
+`Backbone.js`_ and `Marionette`_, let's look at one approach using those
```

Comment:
Since this depends on changes recently introduced, it may be worth mentioning the version we're using?